### DISCUSSION TITLE
Render Phase와 Commit Phase에 대해서 설명해 주세요

### DISCUSSION THREAD

**Discussion by jswith** [woowacourse-study/2022-woowahan-bansanghwe#91] [React]
# Render Phase vs Commit Phase

- 사용자가 만든 React Component -> React Element로 변환 -> 조화 과정 내에서는 fiber에 대응
- React는 브라우저 런타임에서 fiber를 활용하여 Virtual DOM을 생성
- Render Phase와 Commit Phase는 리액트 조화 과정(Conciliation)의 세부 요소.

## Render Phase

- 변화된 사항들을 감지하고, 변화된 사항들이 반영된 alternate virtual DOM tree를 생성, Diffing Algorithm을 통해 변화된 사항들을 빠르게 적용하기 위한 effects list를 생성.

## Commit Phase

- Render Phase에서 생성된 Alternate Virtual DOM Tree와 effects list를 활용하여 real DOM에 반영하는 과정

## effects list

- Render Phase에서 생성된 real DOM 변경사항들을 가지고 있는 list

## 조화과정 살펴보기

- 초기 Vitrual DOM Tree 존재
- state 혹은 props의 변화가 발생
- 변경사항이 발생한 fiber에 대해 flag(dirty check) 진행
  - 변경된 컴포넌트에 대응되는 fiber에 dirty check
- Render phase 시작
  - root로부터 렌더가 필요한 fiber (B)를 탐색
- Render phase 진행
  - 렌더가 필요한 fiber부터 beginWork() 함수를 실행, 새로운 virtual DOM fiber를 생성.
  - 작업이 완료되면 completeWork() 실행
  - 처음 dirty가 발생한 B fiber의 completeWork() 호출이 끝나면 commit phase로 전환
  - Render phase에서 diffing algorithm을 활용해 effects list를 만듦.
- memoized component의 Render Phase
  - 사용자가 입력한 compare Function을 바탕으로 비교 or 자체적인 shallow compare를 수행, 새로 component를 부를지 말지 판단.
  - 비교 결과 True : 기존 fiber를 재활용
  - 비교 결과 False : 새로 Component funciton이나 class를 호출하여 fiber를 다시 만듦.
- Commit phase 시작
  - effects ist에 있는 변경 사항을 real dom에 반영
  - 변경 사항이 없는 fiber의 경우? effects list에 추가되지 않을 뿐 alternate Tree에는 추가됨, 다만 current와 alternate에 있는 fiber는 서로 다른 객체
- memoized component의 Commit phase
  - current tree와 alternate tree의 fiber는 같은 포인터를 공유하며, 실제로 같은 객체
- Commit phase 진행
  - effects list에 있는 변경사항을 모두 진행
  - useEffect 실행
- Render Finished
  - 변경사항 반영을 끝낸뒤 새로운 virtual dom tree를 current tree로 가지고 있음

**Comment by al-bur** [comment]
Render Phase와 Commit Phase에 대해서 설명해주세요

**Comment by al-bur** [comment]
변경된 사항을 어떻게 파악할 까요?

**Comment by al-bur** [comment]
virtual DOM이 다르다는 것을 어떤 기준으로 파악할까요?

**Comment by al-bur** [comment]
재조정이 뭘까요?

**Comment by al-bur** [comment]
key는 뭘까요?

**Comment by al-bur** [comment]
render phase와 commit phase를 컴포넌트의 생명주기 관점에서 설명해주세요

**Comment by al-bur** [comment]
컴포넌트 생명주기 아는대로 설명해주세요

**Comment by JUDONGHYEOK** [comment]
리액트 fiber란?

**Comment by al-bur** [comment]
useEffect는 render phase와 commit phase중 어느 시점에 실행이 될까요?

**Comment by al-bur** [comment]
tree란?

**Comment by al-bur** [comment]
왜 tree 형태로 만들었을까요?

**Comment by al-bur** [comment]
tree 형태의 virtualDOM의 단점

**Comment by al-bur** [comment]
DFS, BFS

**Comment by al-bur** [comment]
메모이제이션을 사용해본 적 있나요?

**Comment by JUDONGHYEOK** [comment]
React.memo를 조화과정의 관점에서 설명해주세요

**Comment by JUDONGHYEOK** [comment]
리렌더링 할 때 파악하는 방법? 개발자 도구 같은것들

**Comment by JUDONGHYEOK** [comment]
그러한 요소들이 렌더페이스와 커밋페이스 둘다 일어나는 것인가?

**Comment by JUDONGHYEOK** [comment]
커밋페이즈는 모든 리렌더링 상황에서 일어나는가?

**Comment by al-bur** [comment]
render phase에서 batch

**Comment by JUDONGHYEOK** [comment]
batch를 하고싶지 않은 경우

**Comment by JUDONGHYEOK** [comment]
React.memo를 사용하지 않고 리렌더링이나 렌더링을 방지하는 방법은?

**Comment by JUDONGHYEOK** [comment]
props란 무엇인가?

**Comment by JUDONGHYEOK** [comment]
리액트에서 state란 무엇인가?

**Comment by JUDONGHYEOK** [comment]
react와 reactDOM의 관계?

**Comment by JUDONGHYEOK** [comment]
diff알고리즘은 react와 reactDOM은 어떤 것에 속하나?

**Comment by JUDONGHYEOK** [comment]
클로저에 대해 설명해주세요

**Comment by JUDONGHYEOK** [comment]
전역변수의 문제점은?
